File: string-hard/827.最大人工岛.java
/*
 * @lc app=leetcode.cn id=827 lang=java
 *
 * [827] 最大人工岛
 */

// @lc code=start

// 75/75 cases passed (220 ms)
// Your runtime beats 25.47 % of java submissions
// Your memory usage beats 7.37 % of java submissions (78.2 MB)
class Solution {
    public int largestIsland(int[][] grid) {
        List<Integer> areaOfIsland = new ArrayList<Integer>(List.of(0, 0));
        // 0 is ocean, 1 is land. Let idx starts from 2 to avoid conflict
        int idx = 2;
        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                areaOfIsland.add(this.area(idx, grid, r, c));
                idx++;
            }
        }
        // int res = 0 is not good. When all cell are land, it will return 0 directly 
        int res = areaOfIsland.stream().max(Integer::compareTo).get();
        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                if (grid[r][c] == 0) {
                    int jointArea = this.jointArea(grid, areaOfIsland, r, c);
                    res = Math.max(res, jointArea);
                }
            }
        }
        return res;
    }

    int area(int idx, int[][] grid, int r, int c) {
        if (!this.inArea(grid, r, c)) {
            return 0;
        }
        if (grid[r][c] != 1) {
            return 0;
        }
        // Every cell of grid is set to index of areaOfIsland representing the island that this cell belongs to.
        grid[r][c] = idx;
        return 1 + area(idx, grid, r + 1, c)
                 + area(idx, grid, r - 1, c)
                 + area(idx, grid, r, c + 1)
                 + area(idx, grid, r, c - 1);
    }

    boolean inArea(int[][] grid, int r, int c) {
        return 0 <= r && r < grid.length && 0 <= c && c < grid[0].length;
    }

    int jointArea(int[][] grid, List<Integer> areaOfIsland, int r, int c) {
        Set<Integer> islands = new HashSet<>();
        addIsland(islands, grid, r + 1, c);
        addIsland(islands, grid, r - 1, c);
        addIsland(islands, grid, r, c + 1);
        addIsland(islands, grid, r, c - 1);
        return 1 + islands.stream().map(x -> areaOfIsland.get(x)).reduce(0, (a, b) -> a + b);
        // int res = 1;
        // for (int idx : islands) {
        //     res += areaOfIsland.get(idx);
        // }
        // return res;
    }

    void addIsland(Set<Integer> islands, int[][] grid,int r, int c) {
        if (this.inArea(grid, r, c)) {
            islands.add(grid[r][c]);
        }
    }
}
// @lc code=end

File: string-hard/827.最大人工岛.cpp
/*
 * @lc app=leetcode.cn id=827 lang=cpp
 *
 * [827] 最大人工岛
 */
#include <numeric>
#include <unordered_map>
#include <vector>

using namespace std;

// @lc code=start
struct UnionFind {
    vector<int> nodes, sizes;
    int maxsize = 0;
    UnionFind(int size) : nodes(size), sizes(size, 1) { iota(nodes.begin(), nodes.end(), 0); }
    int find(int node) {
        if (nodes[node] != node)
            return nodes[node] = find(nodes[node]);
        else
            return nodes[node];
    }
    int merge(int node1, int node2) {
        int f1 = find(node1), f2 = find(node2);
        if (f1 == f2)
            return false;
        if (sizes[f1] > sizes[f2])
            swap(f1, f2);
        maxsize = max(maxsize, sizes[f2] += sizes[f1]);
        nodes[f1] = f2;
        return true;
    }
};

class Solution {
   public:
#define node(x, y) ((x)*n + (y))
    int largestIsland(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        UnionFind uf(m * n);
        int island_maxsize = 0;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++) {
                if (!grid[i][j])
                    continue;
                island_maxsize = 1;
                if (i + 1 < m && grid[i + 1][j])
                    uf.merge(node(i, j), node(i + 1, j));
                if (j + 1 < n && grid[i][j + 1])
                    uf.merge(node(i, j), node(i, j + 1));
            }

        int dirs[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        island_maxsize = max(island_maxsize, uf.maxsize);
        unordered_map<int, int> neighbor;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++) {
                if (grid[i][j])
                    continue;
                for (auto& d : dirs) {
                    int nx = i + d[0], ny = j + d[1];
                    if (nx < 0 || ny < 0 || nx >= m || ny >= n || !grid[nx][ny])
                        continue;
                    int fa = uf.find(node(nx, ny));
                    neighbor[fa] = uf.sizes[fa];
                }
                int sumsize = 1;
                for (auto [com, size] : neighbor)
                    sumsize += size;
                island_maxsize = max(sumsize, island_maxsize);
                neighbor.clear();
            }

        return island_maxsize;
    }
};
// @lc code=end
